src/services/polkit/qml.cpp

```diff
@@ -0,0 +1,428 @@
+#include "qml.hpp"
+#include "listener.hpp"
+#include "session.hpp"
+
+#include <grp.h>
+#include <pwd.h>
+#include <qdir.h>
+#include <qfile.h>
+#include <qtimer.h>
+#include <polkitagent/polkitagent.h>
+#include <polkit/polkit.h>
+
+#include "../../core/logcat.hpp"
+
+namespace {
+QS_LOGGING_CATEGORY(logPolkit, "quickshell.service.polkit");
+}
+
+namespace qs::service::polkit {
+
+static const QString emptyString;
+
+Identity::Identity(
+    id_t id,
+    QString name,
+    QString displayName,
+    QString icon,
+    bool isGroup,
+    PolkitIdentity* polkitIdentity,
+    QObject* parent
+)
+    : QObject(parent)
+    , polkitIdentity(polkitIdentity)
+    , mId(id)
+    , mName(std::move(name))
+    , mDisplayName(std::move(displayName))
+    , mIcon(std::move(icon))
+    , mIsGroup(isGroup) {}
+
+Identity::~Identity() = default;
+
+id_t Identity::id() const { return mId; }
+const QString& Identity::name() const { return mName; }
+const QString& Identity::displayName() const { return mDisplayName; }
+const QString& Identity::icon() const { return mIcon; }
+bool Identity::isGroup() const { return mIsGroup; }
+
+SubMessage::SubMessage(QString text, bool isError, QObject* parent)
+    : QObject(parent)
+    , mText(std::move(text))
+    , mIsError(isError) {}
+
+SubMessage::~SubMessage() = default;
+
+const QString& SubMessage::text() const { return mText; }
+bool SubMessage::isError() const { return mIsError; }
+
+InputRequest::InputRequest(QString message, bool echo, QObject* parent)
+    : QObject(parent)
+    , mMessage(std::move(message))
+    , mEcho(echo) {}
+
+InputRequest::~InputRequest() = default;
+
+const QString& InputRequest::message() const { return mMessage; }
+bool InputRequest::echo() const { return mEcho; }
+
+static std::unordered_map<QString, PolkitAgent*> registeredAgentsByPath {};
+static std::unordered_map<QString, PolkitAgent*> waitingAgentsByPath {};
+
+PolkitAgent::PolkitAgent(QObject* parent): QObject(parent), listener(qs_polkit_agent_new(this)) {}
+
+PolkitAgent::~PolkitAgent() {
+	// First, destroy all but the first request in the queue, so that the cancel
+	// method doesn't start a new one.
```

Comment:
I would think its not unreasonable to handle a request right after cancelling a different one, especially if they're for different programs.

Also, when reloading the graph, preserving the old polkit session may be the most reasonable case, especially if the user wants to do something like edit their auth dialog with a real app running to test with.

Reply:
Handling one request after another is the default behavior in `finishAuthenticationRequest`. The `cancelAuthenticationRequest` method calls `finishAuthenticationRequest` and therefore would start another request immediately.

However, we are in the destructor of the agent here, so there will be no-one to handle the queued requests after the destructor has run. Therefore I want to reject all currently queued requests and not start processing them. To this end, I start by rejecting all requests after the current one, so that when the cancel method runs, it won't have any more to start.

Reply:
As for reloading see my comment below about using an internal singleton. you can wait for the destruction of the last engine generation with a ref to the backing agent to cancel all requests and deregister the agent, or some similar pattern.